backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/projection/impl/AbstractNotNestableElasticsearchProjection.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.search.projection.impl;
+
+import java.util.Set;
+
+import org.hibernate.search.backend.elasticsearch.search.common.impl.ElasticsearchSearchIndexScope;
+import org.hibernate.search.engine.reporting.spi.EventContexts;
+import org.hibernate.search.engine.search.projection.spi.SearchProjectionBuilder;
+
+import com.google.gson.JsonObject;
+
+public abstract class AbstractNotNestableElasticsearchProjection<P> implements ElasticsearchSearchProjection<P> {
+
+	protected final Set<String> indexNames;
+
+	protected AbstractNotNestableElasticsearchProjection(ElasticsearchSearchIndexScope<?> scope) {
+		indexNames = scope.hibernateSearchIndexNames();
+	}
+
+	@Override
+	public final Extractor<?, P> request(JsonObject requestBody, ProjectionRequestContext context) {
+		if ( context.absoluteCurrentFieldPath() != null ) {
+			throw log.cannotUseProjectionInNestedContext(
+					projectionName(),
+					unsupportedNestingHint(),
+					EventContexts.indexSchemaRoot()
+			);
+		}
+		return doRequest( requestBody, context );
+	}
+
+	protected abstract String projectionName();
```

Comment:
For consistency, and since it doesn't make sense to translate this anyway, could you rather use projection keys that you'd add to `org.hibernate.search.engine.search.projection.spi.ProjectionTypeKeys`? We already use such keys in some exception messages.

You can add backend-specific variants if necessary, like we did with e.g. `org.hibernate.search.backend.lucene.search.predicate.impl.LucenePredicateTypeKeys`.